util/reflect.go

```diff
@@ -545,6 +548,108 @@ func forEachFieldInternal(ni *NodeInfo, in, out interface{}, iterFunction FieldI
 	return errs
 }
 
+// ForEachDataField iterates the value supplied and iterates iterFunction for
+// each data tree node found in the function. No schema information is required
+// to perform the iteration.
+func ForEachDataField(value, in, out interface{}, iterFunction FieldIteratorFunc) Errors {
+	if IsValueNil(value) {
+		return nil
+	}
+
+	return forEachDataFieldInternal(&NodeInfo{FieldValue: reflect.ValueOf(value)}, in, out, iterFunction)
+}
+
+func forEachDataFieldInternal(ni *NodeInfo, in, out interface{}, iterFunction FieldIteratorFunc) (errs Errors) {
+	if IsValueNil(ni) {
+		return
+	}
+
+	if IsNilOrInvalidValue(ni.FieldValue) {
+		// Skip any fields that are nil within the data tree, since we
+		// do not need to iterate on them.
+		return
+	}
+
+	// Run the iterator function for this field.
+	errs = AppendErrs(errs, iterFunction(ni, in, out))
+
+	v := ni.FieldValue
+	t := v.Type()
+
+	// Determine whether we need to recurse into the field, or whether it is
+	// a leaf or leaf-list, which are not recursed into when traversing the
+	// data tree.
+	switch {
+	case IsTypeStructPtr(t):
+		// A struct pointer in a GoStruct is a pointer to another container within
+		// the YANG, therefore we dereference the pointer and then recurse. If the
+		// pointer is nil, then we do not need to do this since the data tree branch
+		// is unset in the schema.
+		t = t.Elem()
+		v = v.Elem()
+		fallthrough
+	case IsTypeStruct(t):
+		// Handle non-pointer structs by recursing into each field of the struct.
+		for i := 0; i < t.NumField(); i++ {
+			sf := t.Field(i)
+			nn := &NodeInfo{
+				Parent:      ni,
+				StructField: sf,
+				FieldValue:  reflect.Zero(sf.Type),
+			}
+
+			nn.FieldValue = v.Field(i)
+			ps, err := SchemaPaths(nn.StructField)
+			if err != nil {
+				return NewErrs(err)
+			}
+			// In the case that the field expands to >1 different data tree path,
+			// i.e., SchemaPaths above returns more than one path, then we run
+			// this function for each different path. This ensures that the iterator
+			// function runs for all expansions of the data tree as well as the GoStruct
+			// fields.
+			for _, p := range ps {
+				nn.PathFromParent = p
+				if IsTypeSlice(sf.Type) || IsTypeMap(sf.Type) {
+					nn.PathFromParent = p[0:1]
```

Comment:
Can you help me understand this indexed slice?

Reply:
In the library, we sometimes compress elements out of the data tree that are within the YANG schema when `compress_paths` is set to true to the Go generator binary. This allows us to do things like remove the "config" container which isn't really required for folks just writing config generation from the output structs. We also remove containers that are only there to surround lists, since they are only required in vendor's implementations.

In these compression cases, then we end up with a path that looks like `config/bar` or `interfaces/interface` in the schema annotation. This just ensures that if the path were something like `config/bar/` and we had ended up with a length of greater than 2, then we sanitise the input. 

Strictly, this isn't really required, since we can never generate a three element long path in these cases from the `ygen` library. So it could be removed with no negative impact - it seemed worth being consistent though (`forEachFieldInternal` does this).

Reply:
(optional) Thanks that makes sense, what do you think, does this warrant for an inline comment of some sort?

Reply:
Yep, added.